src/main/java/com/linecorp/armeria/common/ServiceInvocationContext.java

```diff
@@ -284,10 +293,20 @@ private static boolean propagateContextIfNotPresent(ServiceInvocationContext pro
             return false;
         }, () -> {
             setCurrent(propagatedContext);
+            if (propagatedContext.makeContextAwareHandlers != null) {
+                propagatedContext.makeContextAwareHandlers.forEach(MakeContextAwareHandler::enter);
+            }
             return true;
         });
     }
 
+    public void addMakeContextAwareHandler(Runnable onContextEnter, Runnable onContextExit) {
```

Comment:
How about splitting this into two methods such as:

- `onEnter(Runnable)`
- `onExit(Runnable)`

..?

The resulting code would be:

```java
ctx.onEnter(() ->...)
   .onExit(() -> ...);
```